feat: Implement real-time call graph visualization and expansion

caden-peng opened the pull request.

caden-peng committed:
feat: Implement real-time call graph visualization and expansion

caden-peng committed:
fix: resolve duplicate nodes in graph generation. Still need update.

EagleZhen commented on package-lock.json:
why do we need `package-lock.json` at the root directory>

caden-peng committed:
feat: implement dynamic node expansion and visual optimization

caden-peng committed:
merge: resolve conflicts and integrate backend ID logic

EagleZhen commented on experiments/graphdata/src/extension.ts:
why touch the codes in /experiments

caden-peng commented on extension/src/graph.ts:
```diff
@@ -1,72 +1,367 @@
+ * [修复版] 将 VS Code CallHierarchyItem 转换为内部 FunctionCall 类型。
+ * 关键修改：始终使用 item.range (定义位置) 来作为 ID 的一部分，忽略具体的调用行号。
+ * 这确保了同一个函数在图中永远只有一个唯一的节点 ID。
+ */
```
solved

caden-peng replied:
ai suggestions solved in 2f1f506

caden-peng commented on extension/src/graph.ts:
```diff
@@ -1,72 +1,367 @@
+    // [修复] 只传递 item，不传递 range，避免 ID 包含调用位置
+    incoming: rawHierarchy.callers.map((c) => toFuncCall(c.from)),
+    // [修复] 只传递 item，不传递 range
+    outgoing: rawHierarchy.callees.map((c) => toFuncCall(c.to)),
```
solved

caden-peng commented on extension/src/templates/graphScript.js:
```diff
@@ -92,17 +132,78 @@ function initializeCytoscape(graphData) {
+      if (Array.isArray(message.data.edges)) {
+        for (const edge of message.data.edges) {
+          // Check if edge already exists
+          if (cy.getElementById(edge.data.id).empty()) {
```
solved

caden-peng commented on extension/src/graph.ts:
```diff
@@ -1,72 +1,367 @@
+    edges: graph.edges.map((edge) => ({
       data: {
+        // [FIX] This now correctly encodes the RAW edge ID just once
+        id: encodeId(edge.data.id!),
```
solved

caden-peng commented on experiments/graphdata/src/extension.ts:
```diff
@@ -1,161 +1,127 @@
+      // 17.  const targetId = `${hierarchy.target.name} @ ${hierarchy.target.filePath}:${hierarchy.target.line}`;
+      // 18.  const graphData = transformToCytoscapeGraph(hierarchy);
+      // 19.  showGraphView(context, graphData, targetId, output, getCallHierarchyAt);
+      // 20. }
```
solved

caden-peng commented on extension/src/templates/graphScript.js:
```diff
@@ -1,14 +1,29 @@
+ * @param {cytoscape.Collection} nodes - The nodes to check.
+ */
+function updateExpandableNodes(nodes) {
+  // We will implement this later when we have real multi-level data
```
solved

caden-peng commented on extension/src/graph.ts:
```diff
@@ -1,72 +1,367 @@
+  CytoscapeEdge,
+} from "@/types";
+// Import the analyzer function to use its *type*
+import { getCallHierarchyAt } from "@/analyzer";
```
solved

caden-peng commented on extension/src/extension.ts:
```diff
@@ -1,13 +1,30 @@
+  showGraphView,
+  transformToCytoscapeGraph,
+  convertVsCodeHierarchy,
+  toFuncCall, // We'll use this for the debug command
```
solved

caden-peng committed:
fix: resolve merge conflicts and rendering bugs

- Resolved git conflicts in extension.ts and analyzer.ts
- Fixed duplicate function definition in graph.ts
- Added null checks for edge IDs in graphScript.js to prevent rendering errors
- Implemented updateExpandableNodes stub for future use

caden-peng committed:
chore: removed package-lock.json caused by running npm install

caden-peng commented:
https://github.com/user-attachments/assets/4b1fcb35-8af8-44f8-b20c-73aa58e2a343

EagleZhen reviewed (approved):
The overall result seems reasonable. However, as the changes are too large for effective manual review, I did not really investigate the details.

It would be clearer if you could include the commit hash in the reply to those code review comments.
